examples/cosmos/ics20/ics20.qnt

```diff
@@ -0,0 +1,242 @@
+    pure val bankResult = if (movingBackAlongTrace({port: sourcePort, channel: sourceChannel}, denomination)) {
+      // burn vouchers
+      (chainState.bank).BurnCoins(sender, denomination, amount)
+    } else {
+      // escrow tokens
+      pure val escrowAccount = chainState.channelEscrowAddresses.get(sourceChannel)
+      chainState.bank.TransferCoins(sender, escrowAccount, denomination, amount)
+    }
+
+    if (bankResult.success) {
+      pure val data = { denom: denomination, amount: amount, sender: sender, receiver: receiver, memo: "" }
+
+      // handler.sendPacket
+      pure val packet = {
+        data: data,
+        sourcePort: sourcePort,
+        sourceChannel: sourceChannel,
+        destPort: "transfer",
+        destChannel: getCounterparty(chainState, sourceChannel)
+      }
+
+      chainState
+        .with("bank", bankResult.accounts)
+        .with("outPackets", chainState.outPackets.union(Set(packet)))
+    } else {
+      chainState
+    }
+  }
+
+  pure def onRecvPacket(chainState: ChainState, packet: Packet): ReceiveTokenPacketResult = {
+    pure val data = packet.data
+    if (movingBackAlongTrace({port: packet.sourcePort, channel: packet.sourceChannel}, data.denom)) {
+      // unescrow tokens to receiver
+      pure val escrowAccount = chainState.channelEscrowAddresses.get(packet.destChannel)
+      pure val receiverDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
+
+      pure val bankResult = chainState.bank.TransferCoins(escrowAccount, data.receiver, receiverDenom, data.amount)
+
+      pure val newChainState = chainState.with("bank", bankResult.accounts)
+      pure val ack = if (bankResult.success) {
+        { success: true, errorMessage: "" }
+      } else {
+        { success: false, errorMessage: "transfer coins failed" }
+      }
+
+      { chainState: newChainState, acknowledgement: ack }
+    } else {
+      // mint vouchers to receiver
+      pure val newDenom = updateTrace(data.denom, packet.sourcePort, packet.sourceChannel, packet.destPort, packet.destChannel)
+
+      pure val bankResult = chainState.bank.MintCoins(data.receiver, newDenom, data.amount)
+
+      pure val newChainState = chainState.with("bank", bankResult.accounts)
+      pure val ack = if (bankResult.success) {
+        { success: true, errorMessage: "" }
+      } else {
+        { success: false, errorMessage: "mint coins failed" }
+      }
+
+      { chainState: newChainState, acknowledgement: ack }
+    }
+  }
+}
+
+module ics20Test {
+  import base.* from "./base"
+  import ics20.*
+
+  /// Map from chain identifiers to their state
+  var chainStates: str -> ChainState
+
+  /// Map from chain identifiers to their connections, which are maps from
+  /// counterparty chain identifiers to channel identifiers. For example, chain
+  /// A connects to chain B through channel "channelToB".
+  pure val connections = Map(
```

Comment:
I am a bit confused here. The IBC standards have the concept of connection (in the stack between clients and channels), but I guess this is not what is captured here? It seems that this is more abstract an captures the network of chains (the network "topology")? 
If this interpretation is right, I would rename `connections` to `network`?